File: CS/LookUpEditFilterExample/Controls/LookUpEditExt.cs
using DevExpress.Data.Filtering;
using DevExpress.Xpf.Editors.Helpers;
using DevExpress.Xpf.Grid.LookUp;
using System.Linq;
using System.Reflection;

namespace LookUpEditFilterExample.Controls {
    public class LookUpEditExt : LookUpEdit {
        static LookUpEditExt() {
            LookUpEditExtSettings.RegisterCustomEdit();
        }
        public LookUpEditExt() { }
    }

    public class LookUpEditExtSettings : LookUpEditSettings {
        static LookUpEditExtSettings() {
            RegisterCustomEdit();
        }

        protected override IItemsProvider2 CreateItemsProvider() {
            return new ItemsProvider2Ext(this);
        }

        public static void RegisterCustomEdit() {
            EditorSettingsProvider.Default.RegisterUserEditor(typeof(LookUpEditExt), typeof(LookUpEditExtSettings), () => new LookUpEditExt(), () => new LookUpEditExtSettings());
        }
    }

    public class ItemsProvider2Ext : ItemsProvider2, IItemsProvider2 {
        public ItemsProvider2Ext(IItemsProviderOwner owner) : base(owner) { }

        CriteriaOperator IItemsProvider2.CreateDisplayFilterCriteria(string searchText, FilterCondition filterCondition) {
            if (string.IsNullOrEmpty(searchText))
                return null;
            FunctionOperatorType functionOperatorType = GetFunctionOperatorType(filterCondition);
            var filter = new GroupOperator(GroupOperatorType.Or);
            foreach (PropertyInfo pi in ((DevExpress.Data.IListWrapper)DataController.ListSource).WrappedListType.GetGenericArguments().Single().GetProperties())
                filter.Operands.Add(new FunctionOperator(functionOperatorType, new OperandProperty(pi.Name), searchText));
            return filter;
        }
    }
}

File: VB/LookUpEditFilterExample/Controls/LookUpEditExt.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Xpf.Editors.Helpers
Imports DevExpress.Xpf.Grid.LookUp
Imports System.Linq
Imports System.Reflection

Namespace LookUpEditFilterExample.Controls
    Public Class LookUpEditExt
        Inherits LookUpEdit

        Shared Sub New()
            LookUpEditExtSettings.RegisterCustomEdit()
        End Sub
        Public Sub New()
        End Sub
    End Class

    Public Class LookUpEditExtSettings
        Inherits LookUpEditSettings

        Shared Sub New()
            RegisterCustomEdit()
        End Sub

        Protected Overrides Function CreateItemsProvider() As IItemsProvider2
            Return New ItemsProvider2Ext(Me)
        End Function

        Public Shared Sub RegisterCustomEdit()
            EditorSettingsProvider.Default.RegisterUserEditor(GetType(LookUpEditExt), GetType(LookUpEditExtSettings), Function() New LookUpEditExt(), Function() New LookUpEditExtSettings())
        End Sub
    End Class

    Public Class ItemsProvider2Ext
        Inherits ItemsProvider2
        Implements IItemsProvider2

        Public Sub New(ByVal owner As IItemsProviderOwner)
            MyBase.New(owner)
        End Sub

        Private Function IItemsProvider2_CreateDisplayFilterCriteria(ByVal searchText As String, ByVal filterCondition As FilterCondition) As CriteriaOperator Implements IItemsProvider2.CreateDisplayFilterCriteria
            If String.IsNullOrEmpty(searchText) Then
                Return Nothing
            End If
            Dim functionOperatorType As FunctionOperatorType = GetFunctionOperatorType(filterCondition)
            Dim filter = New GroupOperator(GroupOperatorType.Or)
            For Each pi As PropertyInfo In DirectCast(DataController.ListSource, DevExpress.Data.IListWrapper).WrappedListType.GetGenericArguments().Single().GetProperties()
                filter.Operands.Add(New FunctionOperator(functionOperatorType, New OperandProperty(pi.Name), searchText))
            Next pi
            Return filter
        End Function
    End Class
End Namespace
